JP-4241: Use trace model in pixel replace step (#10590)
spacetelescope/jwst: docs/jwst/pixel_replace/main.rst

During 1-D spectral extraction (:ref:`extract_1d <extract_1d_step>` step),
pixels flagged as bad are ignored in the summation process.
If a bad pixel is part of the point-spread function (PSF) at a given wavelength, the
absence of the signal in the flagged pixel will lead to a gap at that wavelength in
the extracted spectrum.

For MIRI MRS and NIRSpec IFU exposures, NaN-valued pixels are partially compensated for during the
IFU cube building process by the overlap between dithered detector pixels and output cube voxels.
The effects of NaN values are thus not as severe as for slit spectra, but can manifest
as small dips in the extracted spectrum when a NaN value lands atop the peak of a spectral

> trace and cube building reconstructs the output voxel from lower-flux adjacent values.

Reviewer: ```suggestion
when a NaN value lands atop the peak of a spectral
trace. In that case, cube building reconstructs the output voxel from lower-flux adjacent values, leading to a small dip in the extracted spectrum 
```
suggestion to fix slight run-on sentence, hopefully I preserved the meaning

> To minimize such artifacts in the 1-D spectra this step uses interpolation methods to estimate the flux values of

Reviewer: ```suggestion
To minimize such artifacts in the 1-D spectra, this step uses interpolation methods to estimate the flux values of
```

missing pixels flagged as ``DO_NOT_USE`` in 2-D calibrated spectra
prior to rectification in the :ref:`resample_spec <resample_spec_step>` or
:ref:`cube_build <cube_build_step>` steps.
The ``pixel_replace`` step inserts these estimates into the 2-D spectral image array,

> unsets the ``DO_NOT_USE`` flag in the DQ array and sets the ``FLUX_ESTIMATED`` flag for

Reviewer: ```suggestion
unsets the ``DO_NOT_USE`` flag in the DQ array, and sets the ``FLUX_ESTIMATED`` flag for
```

each affected pixel instead. Error values and variance components for the replaced pixels
are similarly updated with estimated values, following the same interpolation method as is
used for the data.

> Empirically, this step can significantly reduce the number of artifacts in 1d spectra by using information in

Reviewer: ```suggestion
Empirically, this step can significantly reduce the number of artifacts in 1-D spectra by using information in
```

values for single missing pixels in most circumstances (point sources, extended sources, and emission
features).  However, it cannot estimate values for larger
blocks of bad pixels (such as 3x3 bad pixel regions often seen in NIRSpec data) and hence cannot fix some larger
spectral artifacts that the other two methods can.

Trace Modeling
^^^^^^^^^^^^^^

For some exposure types, it is possible to build a detailed spectral trace model
with the :ref:`adaptive_trace_model step <adaptive_trace_model_step>`.  The trace model
usually includes estimated flux values only for high signal-to-noise regions over the core
of the PSF, but it is generally a very good estimate for those values.
If ``algorithm = "trace_model"``, the ``pixel_replace`` step will preferentially replace
any missing values with estimates from the trace model if present.  Any remaining missing
values will be interpolated with the minimum gradient method, above.

If the :ref:`adaptive_trace_model step <adaptive_trace_model_step>` was not run prior to calling
``pixel_replace``, then it will be run at the start of the step, with the step parameter
``oversample=1`` and otherwise default values.

> The errors and variances associated with pixels replaced from the trace model must still

Reviewer: ```suggestion
The errors and variances associated with pixels replaced from the trace model must also
```

be interpolated, since the trace model does not provide error estimates.  For these pixels,
the error values are linearly interpolated from the nearest valid pixels along the dispersion

> direction.

Reviewer: ```suggestion
be interpolated, since the trace model does not provide error estimates.  These error values are linearly interpolated from the nearest valid pixels along the dispersion
direction.
```
